datasource/ethereum/src/block_ingestor.rs

```diff
@@ -0,0 +1,140 @@
+use failure::Error;
+use std::fmt::Debug;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::time::Instant;
+use web3::api::Web3;
+use web3::transports::batch::Batch;
+use web3::types::Block;
+use web3::types::BlockId;
+use web3::types::BlockNumber;
+use web3::types::H256;
+use web3::types::Transaction;
+use web3::BatchTransport;
+
+use graph::prelude::*;
+
+#[derive(Clone, Debug)]
+pub struct BlockIngestorConfig<S: BlockStore, T: BatchTransport> {
+    pub store: Arc<Mutex<S>>,
+    pub network_name: String,
+    pub web3_transport: T,
+    pub ancestor_count: u64,
+    pub logger: slog::Logger,
+    pub polling_interval: Duration,
+}
+
+pub struct BlockIngestor<S: BlockStore, T: BatchTransport + Debug + Clone> {
+    store: Arc<Mutex<S>>,
+    network_name: String,
+    web3_transport: T,
+    ancestor_count: u64,
+    logger: slog::Logger,
+    polling_interval: Duration,
+}
+
+impl<S: BlockStore + Send + 'static, T: BatchTransport + Debug + Clone> BlockIngestor<S, T> {
+    pub fn spawn(config: BlockIngestorConfig<S, T>) -> Result<(), Error>
+    where
+        T: Send + Sync + 'static,
+    {
+        // Extract config params
+        let BlockIngestorConfig {
+            store,
+            network_name,
+            web3_transport,
+            ancestor_count,
+            logger,
+            polling_interval,
+        } = config;
+
+        // Add a head block pointer for this network name if one does not already exist
+        store.lock().unwrap().add_network_if_missing(&network_name)?;
+
+        // Start block ingestor
+        let block_ingestor = BlockIngestor {
+            store,
+            network_name,
+            web3_transport,
+            ancestor_count,
+            logger: logger.new(o!("component" => "BlockIngestor")),
+            polling_interval,
+        };
+        tokio::spawn(block_ingestor.into_polling_stream());
+
+        Ok(())
+    }
+
+    fn into_polling_stream(self) -> impl Future<Item = (), Error = ()> {
+        let err_logger = self.logger.clone();
+
+        tokio::timer::Interval::new(Instant::now(), self.polling_interval)
+            .map_err(Error::from)
+            .map(move |_| {
+                self.do_poll().unwrap_or_else(|e| {
+                    error!(self.logger, "failed to poll for latest block: {:?}", e);
+                });
+            })
+            .collect()
+            .map(|_| ())
+            .map_err(move |e| {
+                error!(err_logger, "timer::Interval failed: {:?}", e);
+            })
+    }
+
+    fn do_poll(&self) -> Result<(), Error> {
+        let latest_block = self.get_latest_block()?;
+
+        let mut new_blocks = vec![latest_block];
+        while new_blocks.len() > 0 {
+            let missing_block_hashes = self.ingest_blocks(&new_blocks)?;
+            new_blocks = self.get_blocks(&missing_block_hashes)?;
```

Comment:
Generally you did a great job of documenting the algorithm, but this section of the code is scarce on comments and I'm finding it a bit hard to follow. What I imagine is happening here in case of a reorg:

- `latest_block` is ingested, we get back a list of up to `ancestor_count` blocks.
- The ancestors on the list are fetched.
- The ancestors are ingested and `attempt_head_update` is a no-op because `latest_block` remains the head.

If this is the algorithm then this while loop doesn't represent it well. If it's not then I'm confused and we could use some comments explaining when this would iterate more than once and when it would terminate.

Reply:
Yeah it is hard to follow. I added some comments here. Let me know if it makes more sense now